Add Molmo2 multimodal model support

sangho-vision opened the pull request:
## Purpose

This PR adds support for [Molmo2](https://huggingface.co/collections/allenai/molmo2) to vLLM.
Molmo2 is a family of open vision-language models developed by the Allen Institute for AI (Ai2) that support image, video and multi-image understanding and grounding.

This change:
- Introduces a new Molmo2 model implementation for vLLM
- Registers Molmo2 in the multimodal model registry
- Adds a Molmo2-specific video loader that supports video decoding and frame sampling via `decord`, `PyAV`, or `torchcodec`
- Enables inference for image-, multi-image-, and video-conditioned generation
- Updates documentation and examples to demonstrate Molmo2 usage in vLLM

FIX https://github.com/vllm-project/vllm/issues/31331

## Test Plan

```python
import os
import torch
from vllm import LLM
from vllm.sampling_params import SamplingParams

# Please use the `molmo2` video loader backend when passing video urls
os.environ["VLLM_VIDEO_LOADER_BACKEND"] = "molmo2"

model_dir = "allenai/Molmo2-8B"

llm = LLM(
    model=model_dir,
    trust_remote_code=True,
    dtype="bfloat16",
    limit_mm_per_prompt={"image": 2, "video": 1},
    max_num_batched_tokens=36864,  # model max len is 36864
)

sampling_params = SamplingParams(
    max_tokens=2048,
    temperature=0
)

image_messages = [
    {
        "role": "user",
        "content": [
            dict(type="text", text="Compare these images."),
            dict(type="image_url", image_url={"url": "https://picsum.photos/id/237/536/354"}),
            dict(type="image_url", image_url={"url": "https://vllm-public-assets.s3.us-west-2.amazonaws.com/vision_model_images/cherry_blossom.jpg"}),
        ]
    }
]

image_outputs = llm.chat(
    image_messages,
    sampling_params=sampling_params,
    use_tqdm=False,
)

print(f"image output: {image_outputs[0].outputs[0].text}")

video_messages = [
    {
        "role": "user",
        "content": [
            dict(type="text", text="Which animal appears in the video?"),
            dict(type="video_url", video_url={"url": "https://storage.googleapis.com/oe-training-public/demo_videos/many_penguins.mp4"}),
        ]
    }
]

video_outputs = llm.chat(
    video_messages,
    sampling_params=sampling_params,
    use_tqdm=False,
)

print("\n")

print(f"video output: {video_outputs[0].outputs[0].text}")
```

## Test Result

```
image output: These two images present a striking contrast in both subject matter and mood.

The first image captures an intimate, close-up view of a black Labrador puppy. The puppy's glossy coat and expressive eyes are the focal point, set against a rustic wooden background. The warm, sepia-toned filter gives the image a nostalgic, vintage feel. The puppy's direct gaze creates a sense of connection with the viewer, evoking feelings of warmth and innocence.

In contrast, the second image offers a sweeping, wide-angle view of a cityscape dominated by a tall, white tower resembling the Space Needle. The scene is framed by cherry blossom trees in full bloom, creating a beautiful juxtaposition of nature and urban architecture. The bright blue sky and vibrant pink flowers create a lively, cheerful atmosphere.

While the first image focuses on a single, adorable subject in a cozy, intimate setting, the second image presents a grand, expansive view that captures the beauty of both nature and human engineering. The puppy image evokes a sense of warmth and individuality, while the cityscape image inspires awe and appreciation for the harmony between nature and urban development.

Both images are visually striking in their own right, but they convey very different emotions and perspectives on the world around us.

video output: Penguins appear in the video.
```

## Note

Molmo2 uses a **prefix language modeling (prefix LM)** formulation.
As a result, it requires the **TritonAttention-based implementation** recently added to vLLM for correct and efficient execution.

In particular, this PR depends on the following changes:
- https://github.com/vllm-project/vllm/pull/30386
- https://github.com/vllm-project/vllm/pull/30974

These PRs introduce and refine TritonAttention support for prefix LM-style attention patterns, which are necessary for Molmo2 to function correctly in vLLM.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 04e711a7e78ad5f3ef21525e55013ca0863d00a4. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds first-class Molmo2 support across the stack.
> 
> - New `Molmo2ForConditionalGeneration` model with vision backbone, projector, text backbone, tokenizer/processor wrapper, prompt updates, and embedding merge logic (`vllm/model_executor/models/molmo2.py`); registered in `models/registry.py`
> - Marks Molmo2 as prefix LM in `config/model.py` for bidirectional MM attention handling
> - New Molmo2-specific video backend (`vllm/multimodal/video.py`) with OpenCV decoding and frame sampling utilities
> - Processor utilities: add `PromptUpdateDetails.select_token_ids` for multi-token placeholder replacement
> - Examples: add `run_molmo2` (image/video) and `load_molmo2` (multi-image); wire into example model maps; flag `molmo2` as needing video metadata
> - Tests/registry: add HF example info and MM data patches for Molmo2; require Transformers >= 4.51
> - Docs: add Molmo2 to supported multimodal models table
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 04e711a7e78ad5f3ef21525e55013ca0863d00a4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds first-class Molmo2 support across vLLM.
> 
> - New `Molmo2ForConditionalGeneration` model (vision backbone, projector, text stack, HF processor wrapper, prompt updates, multimodal embedding merge) in `vllm/model_executor/models/molmo2.py`; registered in `models/registry.py`
> - Marks Molmo2 as prefix LM in `config/model.py` for correct MM attention behavior
> - New `molmo2` video backend in `vllm/multimodal/video.py` (OpenCV decoding, frame sampling utilities)
> - Processor utility: add `PromptUpdateDetails.select_token_ids` for multi-token placeholder replacement
> - Examples: add `run_molmo2` (image/video) and `load_molmo2` (multi-image); wired into example maps; `molmo2` flagged as needing video metadata
> - Tests/registry: add HF example info and MM data patches; require Transformers >= 4.51
> - Docs: list Molmo2 in supported multimodal models
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit adbc3c6a0841527dc6efb4a40690e28a6a76f532. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces native Molmo2 support across the stack.
> 
> - New `Molmo2ForConditionalGeneration` with vision backbone, projector, text stack, HF processor wrapper, prompt updates, and MM embedding merge (`vllm/model_executor/models/molmo2.py`); registered in `models/registry.py`
> - Marks `molmo2` as prefix-LM for bidirectional MM attention (`config/model.py`); adds `PromptUpdateDetails.select_token_ids` for multi-token placeholder handling
> - Adds `molmo2` video backend with OpenCV decoding and frame sampling utilities (`vllm/multimodal/video.py`)
> - Examples: `run_molmo2` (image/video) and `load_molmo2` (multi-image); wired into model maps; `molmo2` flagged to require video metadata
> - Tests/registry: HF example info (Transformers >= 4.51), MM data patches, and rule to avoid simultaneous image+video in processing tests
> - Docs: list Molmo2 in supported multimodal models
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 449cddf27f957abdfcacdbfcc921707453075649. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 71adcf24a35da619b328e5b77aa9811614b8fb8e. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces first-class Molmo2 support across vLLM.
> 
> - New `Molmo2ForConditionalGeneration` (vision backbone, projector, text stack) with HF processor wrapper and multimodal embedding merge; registered in model registry and marked as prefix LM in `config/model.py`
> - Molmo2-specific video backend (`video.py`) using OpenCV with frame sampling utilities; `MODELS_NEED_VIDEO_METADATA` updated
> - Processing utility: add `PromptUpdateDetails.select_token_ids` for multi-token placeholder replacement
> - Tests/registry: HF example info (Transformers >= 4.51), MM data patches, and rule to avoid simultaneous image+video for Molmo2
> - Examples: `run_molmo2` (image/video) and `load_molmo2` (multi-image); wired into example maps
> - Docs: list Molmo2 in supported multimodal models
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 71adcf24a35da619b328e5b77aa9811614b8fb8e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 67b82e6f8764fd1cb30ea47eed8dc13f5fdf35ab. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces end-to-end Molmo2 support across vLLM.
> 
> - New `Molmo2ForConditionalGeneration` with vision backbone, projector, prefix‑LM text model, processor wrapper, prompt updates, and MM embedding merge (`vllm/model_executor/models/molmo2.py`); registered in models registry
> - Marks `molmo2` as prefix‑LM in `config/model.py`; adds `PromptUpdateDetails.select_token_ids` for multi-token placeholder handling
> - Adds Molmo2-specific video backend in `vllm/multimodal/video.py` (OpenCV decoding, frame sampling); `MODELS_NEED_VIDEO_METADATA` updated
> - Examples: `run_molmo2` (image/video) and `load_molmo2` (multi-image); wired into example maps
> - Tests/registry: HF example info (Transformers >= 4.51), MM data patches, and rule to avoid simultaneous image+video inputs for Molmo2
> - Docs: list Molmo2 in supported multimodal models
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 67b82e6f8764fd1cb30ea47eed8dc13f5fdf35ab. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

sangho-vision committed:
Add Molmo2 multimodal model support

Signed-off-by: sanghol <sanghol@allenai.org>

sangho-vision committed:
Fix bugs in Molmo2 code

Signed-off-by: sanghol <sanghol@allenai.org>

DarkLight1337 commented on examples/offline_inference/vision_language.py:
```diff
@@ -1927,6 +1958,7 @@ def run_tarsier2(questions: list[str], modality: str) -> ModelRequestData:
     "glm4_5v_fp8",
     "qwen3_vl",
     "qwen3_vl_moe",
+    "molmo2",
```
Keep this in alphabetical order

sangho-vision replied:
sorry I updated the code.

DarkLight1337 commented on vllm/model_executor/models/molmo2.py:
```diff
@@ -0,0 +1,2611 @@
+            quant_config=quant_config,
+        )
+        self.scale = self.head_dim**-0.5
+        self.use_pytorch_sdpa = use_pytorch_sdpa
```
cc @Isotr0py is there a difference between this and `TORCH_SDPA` backend?

DarkLight1337 commented on vllm/model_executor/models/molmo2.py:
```diff
@@ -0,0 +1,2611 @@
+            input_dim,
+            self.total_num_heads * self.head_dim,
+            bias=use_bias,
+            quant_config=quant_config,
```
Add `prefix` argument to the attention and feedforward linear layers? (Though I see this isn't done for base Molmo model either...)

DarkLight1337 replied:
Maybe we can simply update the implementation in base Molmo model and import them for Molmo2?

DarkLight1337 replied:
The prefix is used by some quantization methods to find the weights. Basically you just need to match the attribute name; refer to other models for some examples.

cc @Isotr0py

DarkLight1337 reviewed (commented):
Some initial comments

sangho-vision committed:
Update Molmo2 offline inference example code

Signed-off-by: sanghol <sanghol@allenai.org>

DarkLight1337 commented on vllm/model_executor/models/molmo2.py:
```diff
@@ -0,0 +1,2611 @@
+        video_use_col_tokens = processor.processor.video_use_col_tokens
+        use_frame_special_tokens = processor.processor.use_frame_special_tokens
+
+        image_tokens = [
```
Oops I forgot about this. Maybe we can add `PromptUpdateDetails.select_token_ids` which uses `torch.isin` like in `select_text`

sangho-vision replied:
That sounds like a good idea. Would you prefer that I implement this on my side, or is this something you're planning to add in vLLM?

DarkLight1337 replied:
I think it's small enough to include in this PR

sangho-vision committed:
Add  for Molmo2 code

Signed-off-by: sanghol <sanghol@allenai.org>

sangho-vision committed:
Fix pre-commit errors and add prefix for Molmo2

Signed-off-by: sanghol <sanghol@allenai.org>

Isotr0py commented on vllm/multimodal/video.py:
```diff
@@ -267,6 +267,398 @@ def load_bytes(
+        backend = cast(Literal["decord", "torchcodec"], kwargs.pop("backend", "decord"))
+        frame_sample_mode = cast(str | None, kwargs.pop("frame_sample_mode", None))
+        max_fps = cast(int, kwargs.pop("max_fps", 0))
+        sampling_fps = cast(int, kwargs.pop("sampling_fps", 0))
```
Can we use OpenCV to load video from bytes? Because neither `decord` nor `torchcodec` is core dependency in vLLM. You can refer to `load_bytes` from above video backends.

sangho-vision replied:
Molmo2 uses timestamp-based frame sampling by default rather than frame-index-based sampling, so I did not originally include an opencv-based `load_bytes`. However, as you suggested, since `decord` and `torchcodec` are not core dependencies, I've added the opencv-backend function that uses frame-index-based sampling.

Isotr0py commented on vllm/model_executor/models/molmo2.py:
```diff
@@ -0,0 +1,2710 @@
+            self.total_num_kv_heads * self.head_dim,
+            bias=use_bias,
+            quant_config=quant_config,
+        )
```
BTW, we can fuse `qkv_proj` through `QKVParallelLinear`.

sangho-vision replied:
I also followed the Molmo vLLM code for this as well -- are there any advantages to fusing these?

sangho-vision replied:
`inputs_kv` is used for image pooling. I implemented a new module class for image pooling, and merged qkv projections in vit attention and kv projections for image pooling.

Isotr0py commented on vllm/model_executor/models/molmo2.py:
```diff
@@ -0,0 +1,2710 @@
+
+
+def _lowest_multiple(x: int, k: int) -> int:
+    return (x // k) * k
```
We can use `round_down`:
https://github.com/vllm-project/vllm/blob/de08b8f61bd412f2bcc3a5f21da26871e39a9078/vllm/utils/math_utils.py#L30-L32

sangho-vision replied:
I've updated the code.

sangho-vision committed:
Update Molmo2 code

register Molmo2 in

add opencv backend for Molmo2

get image tokens from hf config

Signed-off-by: sanghol <sanghol@allenai.org>

sangho-vision commented on vllm/model_executor/models/molmo2.py:
```diff
@@ -0,0 +1,2710 @@
+FRAME_END_TOKEN = "<frame_end>"
+IM_COL_TOKEN = "<im_col>"
+IMAGE_PROMPT = "<|image|>"
+VIDEO_PROMPT = "<|video|>"
```
I've updated the code.

sangho-vision committed:
Update Molmo2 code

implement a new module class for image pooling

merge q, k, and v projectsion in vit attention

merge k and v projections for image pooling

Signed-off-by: sanghol <sanghol@allenai.org>

sangho-vision commented:
Hi @YounesElhjouji, thanks for your interest in our work.
@DarkLight1337 and @Isotr0py are reviewing my code.
@Isotr0py , I've updated the code following your suggestions.
Please let me know if there’s anything else you’d like me to address.

Isotr0py committed:
clean video backend

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py reviewed (approved):
LGTM now, thanks for your patience!

Isotr0py committed:
Merge remote-tracking branch 'upstream/main' into add-molmo2

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py committed:
fix missing metadata

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py commented:
Seems that mm processor tests are failing because of extra bos/eos token added by tokenizer/processor for text-only/token-only inputs when appling mm cache: https://buildkite.com/vllm/ci/builds/46313#019ba180-7e0a-4494-a7a4-b21ad025dc13

@sangho-vision Can you please take a look?

sangho-vision committed:
Set maximum number of videos to 1

Signed-off-by: sanghol <sanghol@allenai.org>

Isotr0py committed:
no simultaneous MM inputs

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py committed:
Merge branch 'main' into add-molmo2

Isotr0py committed:
fix import

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py committed:
revert and fix bos token

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py committed:
fix

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py committed:
re-insert bos token

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>
